### ISSUE TITLE
Fine tune model by freezing layers

### ISSUE THREAD

**Issue by alisonswu** [Calamari-OCR/calamari#15]
I am training an OCR model for digitalizing historical Arabic documents in Calamari. I wrote an image generator that can produce infinite amount of synthetic images that look similar enough to the original documents. But I only have a very small set of labeled images from the original documents. I would like to explore the idea of training an OCR model using synthetic images, freeze the layers before the LSTM layers, and fine-tuning the model with real data.

I know there are two ways to freeze layers in TensorFlow, by either specifying the trainable variables in the Optimizer
`tf.train.AdamOptimizer(lr).minimize(loss, var_list = list_of_trainable_variables)`
or modifying the option for the individual parameter
`tf.Variable(trainable= False)` 
 I am having trouble figuring out where to apply these changes in Calamari code. Could you give me some suggestions? 

Also, I couldn't find any useful literature on fine-tuning OCR model besides [Tesseract training guide](https://github.com/tesseract-ocr/tesseract/wiki/TrainingTesseract-4.00#training-just-a-few-layers). Do you have suggestion of other methods/literature for fine-tuning OCR models trained on synthetic data with real data?

Thank you!

**Comment by ChWick** [comment]
Calamari does not support freezing of layers (yet?), however pretrained models for finetuning can be used (I use it that way): Load an existing model using the --weights parameter. This will copy the weights of all layers to the new model. The last layer (mapping of LSTM output to probability of each character) is initialized randomly. Some pretrained models can be found here: https://github.com/Calamari-OCR/calamari_models
Note: The network architecture (--network) of the new model must match the structure of the old.

**Comment by alisonswu** [comment]
Update: there is a work-around to freeze the convolutional layers in the default model.
In calamari_ocr/ocr/backends/tensorflow_backend/tensorflow_model.py file, simply add argument **trainable= False** to **create_network()** function.

`layers.append(tf.layers.conv2d(
                            ...,
                            trainable=False,
                   )`

**Comment by christophered** [comment]
@alisonswu can you post a sample of the original document & your synthetic images here?

@ChWick Whats the difference between freezing and continuing from a checkpoint?
Also, what do you think about Alison's proposed freezing additional code?

**Comment by alisonswu** [comment]
@christophered 
Here are some examples from the original document segmented by Tesseract.
![scan1](https://user-images.githubusercontent.com/20031052/44950613-1fc22500-ae1a-11e8-8e07-60d8a513418d.png)
![scan2](https://user-images.githubusercontent.com/20031052/44950614-1fc22500-ae1a-11e8-982d-0f9306bfa389.png)
![scan3](https://user-images.githubusercontent.com/20031052/44950615-1fc22500-ae1a-11e8-9c62-3ca2e36e2849.png)

And examples of synthetic images generated from the same label.
![synthetic1](https://user-images.githubusercontent.com/20031052/44950617-32d4f500-ae1a-11e8-9291-482dc16bc282.png)
![synthetic2](https://user-images.githubusercontent.com/20031052/44950618-32d4f500-ae1a-11e8-90a8-20ed8f4f8aa2.png)
![synthetic3](https://user-images.githubusercontent.com/20031052/44950619-32d4f500-ae1a-11e8-86f7-b7a7eb385cd3.png)
